common/functions/src/scalars/maths/random.rs

```diff
@@ -0,0 +1,73 @@
+// Copyright 2020 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+
+use common_datavalues::prelude::*;
+use common_datavalues::DataSchema;
+use common_datavalues::DataType;
+use common_exception::Result;
+use rand::prelude::*;
+
+use crate::scalars::function_factory::FunctionDescription;
+use crate::scalars::function_factory::FunctionFeatures;
+use crate::scalars::Function;
+
+#[derive(Clone)]
+pub struct RandomFunction {
+    display_name: String,
+}
+
+impl RandomFunction {
+    pub fn try_create(display_name: &str) -> Result<Box<dyn Function>> {
+        Ok(Box::new(RandomFunction {
+            display_name: display_name.to_string(),
+        }))
+    }
+
+    pub fn desc() -> FunctionDescription {
+        FunctionDescription::creator(Box::new(Self::try_create))
+            .features(FunctionFeatures::default())
+    }
+}
+
+impl Function for RandomFunction {
+    fn name(&self) -> &str {
+        &*self.display_name
+    }
+
+    fn num_arguments(&self) -> usize {
+        0
+    }
+
+    fn return_type(&self, _args: &[DataType]) -> Result<DataType> {
+        Ok(DataType::Float64)
+    }
+
+    fn nullable(&self, _input_schema: &DataSchema) -> Result<bool> {
+        Ok(false)
+    }
+
+    fn eval(&self, _columns: &DataColumnsWithField, input_rows: usize) -> Result<DataColumn> {
+        let arr = DFFloat64Array::full(-1.0, input_rows).apply(|_| rand::thread_rng().gen::<f64>());
```

Comment:
Aha, we do not need to create a temp array.

Let's do it with :

```
let iter = (0..input_rows).iter().map(|_| rand::thread_rng().gen::<f64>() );
let array = f64array::new_from_iter(iter);
```

Reply:
`select rand() from numbers(10);` return columns with the same value, not random.

@sundy-li `input_rows` seem always is 0 when call functions with constant parameters

Reply:
It's incorrect, `input_rows` is the row count of the column.

numbers(10) may be divided into N blocks (CPU threads), if you set `max_threads = 1`,  there will be one block with 10 input_rows.

And this could be optimized in constant folding.

Reply:
See `transform_expression_executor.rs`

Reply:
@Junnplus don't understand 

mysql: 
```
 mysql -uroot -e 'WITH DIGITS (N) AS (SELECT 0 UNION ALL SELECT 1 UNION ALL SELECT 2 UNION ALL SELECT 3 UNION ALL   SELECT 4 UNION ALL SELECT 5 UNION ALL SELECT 6 UNION ALL SELECT 7 UNION ALL   SELECT 8 UNION ALL SELECT 9) SELECT RAND() FROM DIGITS'
+---------------------+
| RAND()              |
+---------------------+
|  0.9070275099082175 |
|  0.8253372067821558 |
| 0.40560353491977186 |
|  0.5520060849860368 |
|   0.543219850904513 |
| 0.06008103029809988 |
|  0.6707456295106053 |
| 0.17348508739237217 |
|  0.8551885791636897 |
|   0.755487664374977 |
+---------------------+
```

my random: 
```txt
mysql -h0.0.0.0 -P3307 -uroot -e 'SELECT RAND() from numbers(10)'
+-----------------------+
| RAND()                |
+-----------------------+
| 0.0035617389443762804 |
|    0.3463889841872543 |
|    0.6216070973026121 |
|   0.41610771730421137 |
|   0.06636151314764394 |
|   0.48847821091292687 |
|    0.8531923644472893 |
|    0.3569899290235252 |
|   0.06828498341928513 |
|    0.5615828780217206 |
+-----------------------+
```
imho expected result... please clarify...

Reply:
@mshauneu My fault, local code problem.

Reply:
@Junnplus nw, btw I just updated docs as well.